File: test/financial_calculation_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:yoobbel/controllers/floor_management/marketing_upload_controller.dart';
import 'package:get/get.dart';
import 'test_helper.dart';

void main() {
  setUpAll(() async {
    await TestHelper.setup();
  });

  tearDownAll(() {
    TestHelper.tearDown();
  });

  group('Live Financial Calculations Tests', () {
    late MarketingUploadController controller;

    setUp(() {
      Get.reset();
      controller = Get.put(MarketingUploadController());
    });

    test('Initial product row calculates correct base values', () {
      expect(controller.items.length, 1);
      expect(controller.subTotal.value, 2500.0);
      expect(controller.taxAmount.value, 125.0);
      expect(controller.grandTotal.value, 2625.0);
      expect(controller.balanceDue.value, 2625.0);
    });

    test('Live calculation updates immediately on quantity change', () {
      controller.items[0].quantity.text = '20';
      expect(controller.subTotal.value, 5000.0);
      expect(controller.taxAmount.value, 250.0);
      expect(controller.grandTotal.value, 5250.0);
      expect(controller.balanceDue.value, 5250.0);
    });

    test('Live calculation updates immediately on unit price change', () {
      controller.items[0].quantity.text = '10';
      controller.items[0].orderValue.text = '300';
      expect(controller.subTotal.value, 3000.0);
      expect(controller.taxAmount.value, 150.0);
      expect(controller.grandTotal.value, 3150.0);
    });

    test('Live calculation updates immediately on GST percentage change', () {
      controller.items[0].quantity.text = '10';
      controller.items[0].orderValue.text = '100';
      controller.items[0].gstInfo.text = '18';
      expect(controller.subTotal.value, 1000.0);
      expect(controller.taxAmount.value, 180.0);
      expect(controller.grandTotal.value, 1180.0);
    });

    test('Live calculation updates on advance payment change', () {
      controller.items[0].quantity.text = '10';
      controller.items[0].orderValue.text = '250';
      controller.items[0].gstInfo.text = '5';
      controller.advanceAmount.text = '1000';
      expect(controller.balanceDue.value, 1625.0);

      controller.advanceAmount.text = '2625';
      expect(controller.balanceDue.value, 0.0);
    });

    test('Live calculation updates on shipping charge change', () {
      controller.items[0].quantity.text = '10';
      controller.items[0].orderValue.text = '100';
      controller.items[0].gstInfo.text = '0';
      controller.shippingCharge.text = '150';
      expect(controller.grandTotal.value, 1150.0);
      expect(controller.balanceDue.value, 1150.0);
    });

    test('Adding multiple product items accumulates correctly', () {
      controller.items[0].quantity.text = '10';
      controller.items[0].orderValue.text = '100';
      controller.items[0].gstInfo.text = '5';

      controller.addNewItem();
      expect(controller.items.length, 2);
      expect(controller.subTotal.value, 1000.0 + 2500.0);
      expect(controller.taxAmount.value, 50.0 + 125.0);
      expect(controller.grandTotal.value, 1050.0 + 2625.0);

      controller.items[1].quantity.text = '5';
      controller.items[1].orderValue.text = '200';
      controller.items[1].gstInfo.text = '12';
      expect(controller.subTotal.value, 2000.0);
      expect(controller.taxAmount.value, 170.0);
      expect(controller.grandTotal.value, 2170.0);
    });

    test('Removing product item immediately deducts from totals', () {
      controller.items[0].quantity.text = '10';
      controller.items[0].orderValue.text = '100';
      controller.items[0].gstInfo.text = '0';

      controller.addNewItem();
      controller.items[1].quantity.text = '5';
      controller.items[1].orderValue.text = '200';
      controller.items[1].gstInfo.text = '0';

      expect(controller.subTotal.value, 2000.0);

      controller.removeItem(1);
      expect(controller.items.length, 1);
      expect(controller.subTotal.value, 1000.0);
      expect(controller.grandTotal.value, 1000.0);
    });

    test('Empty or clearing numeric fields safely behaves as zero without crashing', () {
      controller.items[0].quantity.text = '';
      controller.items[0].orderValue.text = '';
      controller.items[0].gstInfo.text = '';
      controller.shippingCharge.text = '';
      controller.advanceAmount.text = '';

      expect(controller.subTotal.value, 0.0);
      expect(controller.taxAmount.value, 0.0);
      expect(controller.grandTotal.value, 0.0);
      expect(controller.balanceDue.value, 0.0);
    });

    test('Decimal values calculate precisely', () {
      controller.items[0].quantity.text = '2.5';
      controller.items[0].orderValue.text = '199.50';
      controller.items[0].gstInfo.text = '18.0';
      controller.shippingCharge.text = '49.99';
      controller.advanceAmount.text = '100.00';

      expect(controller.subTotal.value, closeTo(498.75, 0.001));
      expect(controller.taxAmount.value, closeTo(89.775, 0.001));
      expect(controller.grandTotal.value, closeTo(638.515, 0.001));
      expect(controller.balanceDue.value, closeTo(538.515, 0.001));
    });
  });
}

File: backend/tests/financial_calculation.test.ts
describe('Backend Server-side Financial Calculations', () => {
  it('calculates order line items and grand totals correctly', () => {
    const products = [
      { price: 250, qty: 10, gstPercentage: 5 },
      { price: 150, qty: 4, gstPercentage: 12 },
    ];
    const shipping = 100;
    const advance = 500;

    let subTotal = 0;
    let totalTax = 0;
    let totalQty = 0;

    products.forEach((prod) => {
      const base = prod.price * prod.qty;
      const tax = base * (prod.gstPercentage / 100);

      subTotal += base;
      totalTax += tax;
      totalQty += prod.qty;
    });

    const grandTotal = subTotal + totalTax + shipping;
    const balanceDue = grandTotal - advance;

    // Item 1: 2500 base + 125 tax = 2625
    // Item 2: 600 base + 72 tax = 672
    // Subtotal: 3100
    // Tax: 197
    // Grand: 3100 + 197 + 100 = 3397
    // Balance Due: 3397 - 500 = 2897
    expect(subTotal).toBe(3100);
    expect(totalTax).toBe(197);
    expect(totalQty).toBe(14);
    expect(grandTotal).toBe(3397);
    expect(balanceDue).toBe(2897);
  });

  it('safely handles empty/missing values with zero fallbacks', () => {
    const products = [{ price: undefined, qty: undefined, gstPercentage: undefined }];
    let subTotal = 0;
    let totalTax = 0;
    let totalQty = 0;

    products.forEach((prod: any) => {
      const price = Number(prod.price) || 0;
      const qty = Number(prod.qty) || 1;
      const gstPct = Number(prod.gstPercentage) || 0;

      const base = price * qty;
      const tax = base * (gstPct / 100);

      subTotal += base;
      totalTax += tax;
      totalQty += qty;
    });

    const grandTotal = subTotal + totalTax;
    expect(subTotal).toBe(0);
    expect(totalTax).toBe(0);
    expect(grandTotal).toBe(0);
  });
});
